sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -352,11 +374,16 @@ object TypeCoercion {
         p.makeCopy(Array(Cast(left, TimestampType), right))
       case p @ Equality(left @ TimestampType(), right @ StringType()) =>
         p.makeCopy(Array(left, Cast(right, TimestampType)))
-
       case p @ BinaryComparison(left, right)
-        if findCommonTypeForBinaryComparison(left.dataType, right.dataType).isDefined =>
+        if !plan.conf.binaryComparisonCompatibleWithHive &&
+          findCommonTypeForBinaryComparison(left.dataType, right.dataType).isDefined =>
         val commonType = findCommonTypeForBinaryComparison(left.dataType, right.dataType).get
         p.makeCopy(Array(castExpr(left, commonType), castExpr(right, commonType)))
+      case p @ BinaryComparison(left, right)
+        if plan.conf.binaryComparisonCompatibleWithHive &&
```

Comment:
This is hard to maintain and debug. Instead of mixing them together, could you separate it from the current rule Spark uses?

Reply:
It needs 2 `PromoteStrings` and 2 `InConversion` if separate from current rule, So I merge the logic to `findCommonTypeForBinaryComparison`.

Reply:
I am fine to have two separate rules, but we can call the shared functions.